File: chat-demo-android/src/main/java/com/twilio/chat/demo/BasicChatClient.java
package com.twilio.chat.demo;

import java.util.Arrays;
import java.util.List;

import com.twilio.accessmanager.AccessManager;

import com.twilio.chat.Channel;
import com.twilio.chat.StatusListener;
import com.twilio.chat.CallbackListener;
import com.twilio.chat.ChatClientListener;
import com.twilio.chat.ChatClient;
import com.twilio.chat.ErrorInfo;
import com.twilio.chat.UserInfo;
import com.twilio.chat.internal.HandlerUtil;

import android.app.PendingIntent;
import android.content.Context;
import android.content.Intent;
import android.os.AsyncTask;
import android.os.Handler;
import android.os.Looper;
import android.util.Log;
import android.widget.Toast;

public class BasicChatClient extends CallbackListener<ChatClient>
    implements AccessManager.Listener, AccessManager.TokenUpdateListener
{
    private static final Logger logger = Logger.getLogger(BasicChatClient.class);

    private String accessToken;
    private String fcmToken;

    private ChatClient chatClient;

    private Context       context;
    private AccessManager accessManager;

    private LoginListener loginListener;
    private Handler       loginListenerHandler;

    private String        urlString;
    private String        username;

    public BasicChatClient(Context context)
    {
        super();
        this.context = context;

        if (BuildConfig.DEBUG) {
            ChatClient.setLogLevel(android.util.Log.DEBUG);
        } else {
            ChatClient.setLogLevel(android.util.Log.ERROR);
        }
    }

    public interface LoginListener {
        public void onLoginStarted();

        public void onLoginFinished();

        public void onLoginError(String errorMessage);

        public void onLogoutFinished();
    }

    public void setFCMToken(String fcmToken)
    {
        logger.e("setFCMToken "+fcmToken);
        this.fcmToken = fcmToken;
        if (chatClient != null) {
            setupFcmToken();
        }
    }

    public void login(final String username, final String url, final LoginListener listener) {
        if (username == this.username && urlString == url && loginListener == listener && chatClient != null && accessManager != null) {
            onSuccess(chatClient);
            return;
        }

        this.username = username;
        urlString = url;

        loginListenerHandler = HandlerUtil.setupListenerHandler();
        loginListener = listener;

        new GetAccessTokenAsyncTask().execute(username, urlString);
    }

    public ChatClient getChatClient()
    {
        return chatClient;
    }

    private void setupFcmToken()
    {
        chatClient.registerFCMToken(fcmToken,
            new ToastStatusListener(
                "Firebase Messaging registration successful",
                "Firebase Messaging registration not successful"));
    }

    public void unregisterFcmToken()
    {
        chatClient.unregisterFCMToken(fcmToken,
            new ToastStatusListener(
                "Firebase Messaging unregistration successful",
                "Firebase Messaging unregistration not successful"));
    }

    private void createAccessManager()
    {
        if (accessManager != null) return;

        accessManager = new AccessManager(accessToken, this);
        accessManager.addTokenUpdateListener(this);
    }

    private void createClient()
    {
        if (chatClient != null) return;

        ChatClient.Properties props =
            new ChatClient.Properties.Builder()
                .setRegion("us1")
                .createProperties();

        ChatClient.create(context.getApplicationContext(),
                                 accessToken,
                                 props,
                                 this);
    }

    public void shutdown()
    {
        chatClient.shutdown();
        chatClient = null; // Client no longer usable after shutdown()
    }

    // Client created, remember the reference and set up UI
    @Override
    public void onSuccess(ChatClient client)
    {
        logger.d("Received completely initialized ChatClient");
        chatClient = client;

        if (fcmToken != null) {
            setupFcmToken();
        }

        loginListenerHandler.post(new Runnable() {
            @Override
            public void run()
            {
            if (loginListener != null) {
                loginListener.onLoginFinished();
            }
            }
        });
    }

    // Client not created, fail
    @Override
    public void onError(final ErrorInfo errorInfo)
    {
        TwilioApplication.get().logErrorInfo("Received onError event", errorInfo);

        loginListenerHandler.post(new Runnable() {
            @Override
            public void run()
            {
            if (loginListener != null) {
                loginListener.onLoginError(errorInfo.getErrorCode() + " " + errorInfo.getErrorText());
            }
            }
        });
    }

    // AccessManager.Listener

    @Override
    public void onTokenWillExpire(AccessManager accessManager)
    {
        TwilioApplication.get().showToast("AccessManager.onTokenWillExpire");
    }

    @Override
    public void onTokenExpired(AccessManager accessManager)
    {
        TwilioApplication.get().showToast("Token expired. Getting new token.");
        new GetAccessTokenAsyncTask().execute(username, urlString);
    }

    @Override
    public void onError(AccessManager accessManager, String err)
    {
        TwilioApplication.get().showToast("AccessManager error: " + err);
    }

    // AccessManager.TokenUpdateListener

    @Override
    public void onTokenUpdated(String token)
    {
        TwilioApplication.get().showToast("AccessManager token updated: " + token);

        if (chatClient == null) return;

        chatClient.updateToken(token, new ToastStatusListener(
            "Client Update Token was successfull",
            "Client Update Token failed"));
    }

    /**
     * Modify this method if you need to provide more information to your Access Token Service.
     */
    private class GetAccessTokenAsyncTask extends AsyncTask<String, Void, String>
    {
        @Override
        protected void onPreExecute()
        {
            super.onPreExecute();
            loginListenerHandler.post(new Runnable() {
                @Override
                public void run()
                {
                    if (loginListener != null) {
                        loginListener.onLoginStarted();
                    }
                }
            });
        }

        @Override
        protected String doInBackground(String... params)
        {
            try {
                accessToken = HttpHelper.httpGet(params[0], params[1]);
            } catch (Exception e) {
                e.printStackTrace();
            }
            return accessToken;
        }

        @Override
        protected void onPostExecute(String result)
        {
            super.onPostExecute(result);
            createAccessManager();
            createClient();
            accessManager.updateToken(accessToken);
        }
    }
}

File: chat-demo-android/src/main/java/com/twilio/chat/demo/BasicChatClient.kt
package com.twilio.chat.demo

import com.twilio.chat.internal.HandlerUtil
import android.content.Context
import android.os.AsyncTask
import android.os.Handler
import ToastStatusListener
import com.twilio.chat.*
import org.jetbrains.anko.*
import java.util.Optional

class BasicChatClient(private val context: Context) : CallbackListener<ChatClient>(), ChatClientListener, AnkoLogger {
    private var accessToken: String? = null
    private var fcmToken: String? = null

    var chatClient: ChatClient? = null
        private set

    private var loginListener: LoginListener? = null
    private var loginListenerHandler: Handler? = null

    private var urlString: String? = null
    private var username: String? = null
    private var pinCerts: Boolean = true
    private var realm: String? = null

    init {
        if (BuildConfig.DEBUG) {
            warn { "Enabling DEBUG logging" }
            ChatClient.setLogLevel(android.util.Log.VERBOSE)
        }
    }

    interface LoginListener {
        fun onLoginStarted()
        fun onLoginFinished()
        fun onLoginError(errorMessage: String)
        fun onLogoutFinished()
    }

    private fun notifyLoginStarted() { // Called before getting access token
        loginListenerHandler!!.post {
            if (loginListener != null) {
                loginListener!!.onLoginStarted()
            }
        }
    }
    private fun notifyLoginFinished() { // Called after successful creation of ChatClient
        loginListenerHandler!!.post {
            if (loginListener != null) {
                loginListener!!.onLoginFinished()
            }
        }
    }
    private fun notifyLoginError(errorMessage: String) {
        loginListenerHandler!!.post {
            if (loginListener != null) {
                loginListener!!.onLoginError(errorMessage)
            }
        }
    }
    private fun notifyLogoutFinished() {
        loginListenerHandler!!.post {
            if (loginListener != null) {
                loginListener!!.onLogoutFinished()
            }
        }
    }

    fun setFCMToken(fcmToken: String) {
        warn { "setFCMToken $fcmToken" }
        this.fcmToken = fcmToken
        if (chatClient != null) {
            setupFcmToken()
        }
    }

    fun login(username: String, pinCerts: Boolean, realm: String, url: String, listener: LoginListener) {
        /*
        if (username == this.username
                && pinCerts == this.pinCerts
                && realm == this.realm
                && url == this.urlString
                && listener === loginListener
                && chatClient != null) {
            onSuccess(chatClient!!)
            return
        }
         */
        assert(chatClient == null) { "ChatClient object is to be created on login, should be null before login" }

        this.username = username
        this.pinCerts = pinCerts
        this.realm = realm
        urlString = url

        loginListenerHandler = HandlerUtil.setupListenerHandler()
        loginListener = listener

        getAccessToken()
    }

    fun updateToken() {
        getAccessToken()
    }

    private fun setupFcmToken() {
        chatClient!!.registerFCMToken(fcmToken,
                ToastStatusListener(
                        "Firebase Messaging registration successful",
                        "Firebase Messaging registration not successful"))
    }

    fun unregisterFcmToken() {
        chatClient!!.unregisterFCMToken(fcmToken,
                ToastStatusListener(
                        "Firebase Messaging unregistration successful",
                        "Firebase Messaging unregistration not successful"))
    }

    private fun createClient() {
        assert(chatClient == null)

        val props = ChatClient.Properties.Builder()
                .setRegion(realm)
                .setDeferCertificateTrustToPlatform(!pinCerts)
                .createProperties()

        ChatClient.create(context.applicationContext,
                accessToken!!,
                props,
                this)
    }

    fun shutdown() {
        chatClient!!.shutdown()
        chatClient = null // Client no longer usable after shutdown()
        notifyLogoutFinished()
    }

    // Client created, remember the reference and set up UI
    override fun onSuccess(client: ChatClient) {
        debug { "Received completely initialized ChatClient" }
        chatClient = client

        if (fcmToken != null) {
            setupFcmToken()
        }

        notifyLoginFinished()
    }

    // Client not created, fail
    override fun onError(errorInfo: ErrorInfo?) {
        TwilioApplication.instance.logErrorInfo("Login error", errorInfo!!)
        chatClient = null

        notifyLoginError(errorInfo.toString())
    }

    // Token expiration events

    override fun onTokenAboutToExpire() {
        if (chatClient != null) {
            TwilioApplication.instance.showToast("Token will expire in 3 minutes. Getting new token.")
            getAccessToken()
        }
    }

    override fun onTokenExpired() {
        accessToken = null
        if (chatClient != null) {
            TwilioApplication.instance.showToast("Token expired. Getting new token.")
            getAccessToken()
        }
    }

    private fun getAccessToken() {
        GetAccessTokenAsyncTask().execute(urlString)
    }

    /**
     * Modify this method if you need to provide more information to your Access Token Service.
     */
    //TODO coroutines
    private inner class GetAccessTokenAsyncTask : AsyncTask<String, Void, Optional<String>>() {
        override fun onPreExecute() {
            super.onPreExecute()
            if (chatClient == null) {
                notifyLoginStarted()
            }
        }

        override fun doInBackground(vararg params: String): Optional<String> {
            var result: Optional<String> = Optional.empty();
            try {
                result = Optional.of(HttpHelper.httpGet(params[0]))
            } catch (e: Exception) {
                System.err.println("getAccessToken() error:")
                e.printStackTrace()
                notifyLoginError(e.message.orEmpty())
            }

            return result
        }

        override fun onPostExecute(result: Optional<String>) {
            if (!result.isPresent) return

            accessToken = result.get();

            super.onPostExecute(result)

            applyAccessToken()
        }

        private fun applyAccessToken() {
            if (chatClient == null) {
                // Create client with accessToken
                createClient()
            } else {
                // Client already exists, so set accessToken to it
                chatClient!!.updateToken(accessToken, ToastStatusListener(
                        "Client Update Token was successfull",
                        "Client Update Token failed"))
            }
        }
    }

    override fun onChannelAdded(p0: Channel?) {}
    override fun onChannelDeleted(p0: Channel?) {}
    override fun onChannelInvited(p0: Channel?) {}
    override fun onChannelJoined(p0: Channel?) {}
    override fun onChannelSynchronizationChange(p0: Channel?) {}
    override fun onChannelUpdated(p0: Channel?, p1: Channel.UpdateReason?) {}
    override fun onClientSynchronization(p0: ChatClient.SynchronizationStatus?) {}
    override fun onConnectionStateChange(p0: ChatClient.ConnectionState?) {}
    override fun onRemovedFromChannelNotification(p0: String?) {}
    override fun onUserSubscribed(p0: User?) {}
    override fun onUserUnsubscribed(p0: User?) {}
    override fun onUserUpdated(p0: User?, p1: User.UpdateReason?) {}
    override fun onAddedToChannelNotification(p0: String?) {}
    override fun onInvitedToChannelNotification(p0: String?) {}
    override fun onNewMessageNotification(p0: String?, p1: String?, p2: Long) {}
    override fun onNotificationFailed(p0: ErrorInfo?) {}
    override fun onNotificationSubscribed() {}
}
